### PR TITLE
/review

### LINKED ISSUE #746: AI: `/review`
It would be useful to get ChatGPT's opinion on a pull request by using a `/review` command.
- The `/review` capability should read the original issue specification, and see the code diff (raw diff should be easy to pass in.)
- The prompt should be something along the lines of "I need your help to review a pull request. The original issue specification is XXX and the code diff is XXX. Does this appear to achieve the specification? Your reply will be directly posted on the GitHub pull request and is intended to be read by the assignee and entire review team. Please be descriptive and list all of the actionable changes necessary to achieve the specification if it does not appear to achieve it already."
- This command should be disabled on issue views, and only work on pull request views. The error message should describe this.

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #746 

invoked via /review

The implemented prompts differ from the original spec but after experimenting with variations, I believe the current setup provides uniform formatting and covers all requirements of the spec.

### CODE REVIEW

File: src/handlers/comment/handlers/index.ts [hunk 1]
```diff
@@ -246,6 +248,12 @@ export const userCommands = (): UserCommands[] => {
       handler: query,
       callback: commandCallback,
     },
+    {
+      id: IssueCommentCommands.REVIEW,
+      description: `Ubiquity AI will perform a review of the current pull request. \n  example usage: /review`,
```
[the remark this answers is not in the document]
Comment (pavlovcik) [reply src/handlers/comment/handlers/index.ts:253]: Perhaps an error should be thrown if an associated issue is not found

File: src/handlers/comment/handlers/review.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
+import { getBotContext, getLogger } from "../../../bindings";
+import { GPTResponse, Payload, StreamlinedComment, UserType } from "../../../types";
+import { getAllIssueComments, getAllLinkedIssuesAndPullsInBody, getPullByNumber } from "../../../helpers";
+import { CreateChatCompletionRequestMessage } from "openai/resources/chat";
+import { askGPT, decideContextGPT, gptContextTemplate, specCheckTemplate, speckCheckResponse } from "../../../helpers/gpt";
+
+/**
+ * @notice Three calls to Chad are made. First for context, second for review and third for finalization.
```
Comment (pavlovcik) [inline src/handlers/comment/handlers/review.ts:8]: Chad should be renamed to OpenAI

File: src/handlers/comment/handlers/review.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    logger.info(`Error getting issue comments`);
+    return `Error getting issue comments`;
+  }
+  streamlined.push({
+    login: issue.user.login,
+    body: issue.body,
+  });
+
+  comments.forEach(async (comment, i) => {
+    if (comment.user.type == UserType.User || commentsRaw[i].body.includes("<!--- { 'UbiquityAI': 'answer' } --->")) {
```
Comment (pavlovcik) [inline src/handlers/comment/handlers/review.ts:57]: We don't have any branding related to UbiquityAI. 

I think we can use OpenAI for now in place of these, or ChatGPT (but it would be useful to also note what version is being used if we decide to use "ChatGPT" in place of these)

File: src/handlers/comment/handlers/review.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+      });
+    }
+  });
+
+  // returns the conversational context from all linked issues and prs
+  const links = await getAllLinkedIssuesAndPullsInBody(issue.number);
+
+  if (typeof links === "string") {
+    logger.info(`Error getting linked issues or prs: ${links}`);
+    return `Error getting linked issues or prs: ${links}`;
```
Comment (pavlovcik) [inline src/handlers/comment/handlers/review.ts:70]: Cool to see the error. I believe we are using an error render function now for comments to display as errors.

File: src/handlers/comment/handlers/review.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        })
+        .catch((error) => {
+          logger.info(`Error getting diff: ${error}`);
+          return `Error getting diff: ${error}`;
+        });
+
+      chatHistory.push({
+        role: "system",
+        content: gptContextTemplate,
+        name: "UbiquityAI",
```
Comment (pavlovcik) [inline src/handlers/comment/handlers/review.ts:116]: What is the name field for

File: src/handlers/comment/handlers/review.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+          name: "UbiquityAI",
+        } as CreateChatCompletionRequestMessage,
+        {
+          role: "user",
+          content: "PR Diff: \n" + JSON.stringify(diff), // provide the diff
+          name: "user",
+        } as CreateChatCompletionRequestMessage
+      );
+
+      const draftReport: GPTResponse | string = await askGPT("", chatHistory);
```
Comment (pavlovcik) [inline src/handlers/comment/handlers/review.ts:145]: Seems unusual to pass in an empty string

File: src/helpers/gpt.ts [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+## Overall Analysis
+- [x] {item}
+- [ ] {item}
+`;
+
+export const specCheckTemplate = `
+  You are the UbiquityAI, designed to analyze pull requests in comparison to the issue specification.\n
+  You are to update the report after you review everything, pr diff, spec, and provided context.\n
+  Measure the implementation against the issue spec and provide an overall analysis.\n
+  You must NOT return the any example text, only your comprehensive report. \n
```
Comment (pavlovcik) [inline src/helpers/gpt.ts:50]:
```suggestion
  You must NOT return any example text, only your comprehensive report. \n
```

**Review by 0x4007 — CHANGES_REQUESTED** [review]
Interesting prompts. Seems like a lot of experimentation was put into them. Understanding purely via code review is tough for that but I would love to see QA examples to see those prompts in action.

File: src/handlers/comment/handlers/review.ts (hunk 6 above)
Comment (Keyrxng) [reply src/handlers/comment/handlers/review.ts:145]: The first arg is used only for logging purposes as it's the chatHistory which matters but I'll pass an identifier for logs.

File: src/handlers/comment/handlers/review.ts (hunk 5 above)
Comment (Keyrxng) [reply src/handlers/comment/handlers/review.ts:116]: I'll remove them since we are not using functions.

```ts
export interface ChatCompletionMessageParam {
  /**
   * The contents of the message. `content` is required for all messages, and may be
   * null for assistant messages with function calls.
   */
  content: string | null;

  /**
   * The role of the messages author. One of `system`, `user`, `assistant`, or
   * `function`.
   */
  role: 'system' | 'user' | 'assistant' | 'function';

  /**
   * The name and arguments of a function that should be called, as generated by the
   * model.
   */
  function_call?: ChatCompletionMessageParam.FunctionCall;

  /**
   * The name of the author of this message. `name` is required if role is
   * `function`, and it should be the name of the function whose response is in the
   * `content`. May contain a-z, A-Z, 0-9, and underscores, with a maximum length of
   * 64 characters.
   */
  name?: string;
}
```

**Comment by Keyrxng** [comment]
> Interesting prompts. Seems like a lot of experimentation was put into them. Understanding purely via code review is tough for that but I would love to see QA examples to see those prompts in action.

I've deleted the majority from https://github.com/Keyrxng/UbiquityBotTesting/pull/41#issuecomment-1712961314 especially the really terrible responses but you can still sort of the see the progression. 

If this is mandatory I'll spam my issue with various versions and I'll comment the prompt scheme I used for it below, just lmk

**Comment by pavlovcik** [comment]
I just need to see the latest when it is ready for review. Posting all should be fine

**Comment by Keyrxng** [comment]
> I just need to see the latest when it is ready for review. Posting all and highlighting favorites should be fine

https://github.com/Keyrxng/UbiquityBotTesting/pull/41#issuecomment-1713538151

Added some examples using the spec prompt and showing you what the context call response object looks like

File: src/helpers/gpt.ts [hunk 8]
```diff
@@ -46,22 +46,12 @@ export const issueSpecExample = `
...
-  ${issueSpecExample}
-  `;
-
-export const speckCheckResponse = `
-Finalize the report by proofreading the report for any errors.
-Improve the report by adding any additional information you deem relevant.
-Remove anything that is not relevant to the report, if you can replace it with something more relevant, do so.
-
-All replies MUST end with "\n<!--- { 'OpenAI': 'specCheck' } --->".
+Using the provided context, ensure you clearly understand the specification of the issue.
```
Comment (pavlovcik) [inline src/helpers/gpt.ts:49]: This is a much more streamlined prompt. Thanks!

Comment (Keyrxng) [reply src/helpers/gpt.ts:49]: Ah yeah I'm no longer having to check for previous reports, will have it removed.

The responses are throwing a lot of false positives and occasionally are far too anal about the spec so it still needs worked I think. Things like changes to the .gitignore or not being mentioned in the spec for example silly things like that, will try to find a way to have the second call filter these sorts of things.

Comment (pavlovcik) [reply src/helpers/gpt.ts:49]: I don't think I see an issue with that. Extraneous changes are not desirable. It's like when assignees add extra dependencies and don't use them in the code. I would ding them in the review for that and ask them to remove it for example.

File: src/types/shared.ts [hunk 9]
```diff
@@ -1,3 +1,17 @@
 import { Type } from "@sinclair/typebox";
 
 export const TURL = Type.String({ format: "ipv4" });
+
+export interface StreamlinedComment {
+  login: string | null | undefined;
```
Comment (0xcodercrane) [inline src/types/shared.ts:6]: Would you follow up other schema/type definitions for StreamlinedComment and GPTResponse.

**Review by cuddle-dev — CHANGES_REQUESTED** [review]
Reviewed the codebase. The overall feedback is good. Left a comment.

Going through QAs.

File: src/types/shared.ts (hunk 9 above)
Comment (Keyrxng) [reply src/types/shared.ts:6]: no worries and cheers

File: src/helpers/gpt.ts [hunk 10]
```diff
@@ -194,7 +254,7 @@ export const askGPT = async (question: string, chatHistory: CreateChatCompletion
     messages: chatHistory,
     model: "gpt-3.5-turbo-16k",
     max_tokens: config.review.tokenLimit,
-    temperature: 0,
+    temperature: 0.1,
```
Comment (pavlovcik) [inline src/helpers/gpt.ts:257]: This should be `0` why did you change it?

File: src/handlers/comment/commands.ts [hunk 11]
```diff
@@ -6,6 +6,7 @@ export enum IssueCommentCommands {
   PAYOUT = "/payout", // request permit payout
   MULTIPLIER = "/multiplier", // set bounty multiplier (for treasury)
   QUERY = "/query",
+  REVIEW = "/review", // GPT will review the pull request
```
Comment (pavlovcik) [inline src/handlers/comment/commands.ts:9]:
```suggestion
  REVIEW = "/review", // gpt-3.5-turbo-16k will review the pull request
```

Comment (pavlovcik) [reply src/handlers/comment/commands.ts:9]: Why did you pick that model?

File: src/handlers/comment/handlers/review.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+          return ErrorDiff(`No answer found for issue #${issue.number}`);
+        }
+      }
+    } else {
+      return ErrorDiff(`No PR found for issue #${issue.number}`);
+    }
+  };
+
+  const res = await isPull();
+  return res + `\n> <small> Ensure the pull request requirements are in the linked issue's first comment and update it if the scope evolves. </small>.`;
```
Comment (pavlovcik) [inline src/handlers/comment/handlers/review.ts:122]: What is `\n>`

File: src/helpers/issue.ts [hunk 13]
```diff
@@ -697,7 +697,7 @@ export const getAllLinkedIssuesAndPullsInBody = async (issueNumber: number) => {
   const linkedPRStreamlined: StreamlinedComment[] = [];
   const linkedIssueStreamlined: StreamlinedComment[] = [];
 
-  const regex = /https:\/\/github\.com\/[^/\s]+\/[^/\s]+\/(issues|pull)\/(\d+)/gi;
+  const regex = /(#(\d+)|https:\/\/github\.com\/[^/\s]+\/[^/\s]+\/(issues|pull)\/(\d+))/gi;
```
[the remark this answers is not in the document]
Comment (Keyrxng) [reply src/helpers/issue.ts:700]: Well ``getAllLinkedIssuesAndPullsInBody()`` is used in ``decideContextGPT()`` and it benefits from it, and it is used in ``/ask``.

Consider a scenario where the issue opened links a pr, this is able to fetch context from within that PR.

File: src/handlers/comment/handlers/review.ts (hunk 12 above)
Comment (Keyrxng) [reply src/handlers/comment/handlers/review.ts:122]: \n ensures a new line and '>' begins the blockquote

File: src/helpers/gpt.ts [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+`;
+
+export const specCheckTemplate = `
+Using the provided context, ensure you clearly understand the specification of the issue.
+If you do not understand the specification, infer it from the changes made.
+
+You are the first line of defense to ensure the specification is met before reviewers do their job.
+If the specification is not met, you are to provide a clear and concise reason why.
+
+Your response will be posted as a GitHub comment for everyone see so try to compliment the GitHub UI.
```
Comment (pavlovcik) [inline src/helpers/gpt.ts:24]:
```suggestion
Your response will be posted as a GitHub comment for everyone see in the pull request review conversation. 
```

Perhaps this is why it was laying on the compliments so thick.

Comment (Keyrxng) [reply src/helpers/gpt.ts:24]: hahahaha aye very good 🤣

**Review by 0x4007 — COMMENTED** [review]
Home stretch. Be sure to not resolve conversations unless you solve the problems.

File: src/helpers/gpt.ts (hunk 14 above)
Comment (pavlovcik) [reply src/helpers/gpt.ts:24]:
```suggestion
Your response will be posted as a GitHub comment for everyone to see in the pull request review conversation. 
```

File: ubiquibot-config-default.json [hunk 15]
```diff
@@ -8,6 +8,8 @@
   "disable-analytics": false,
   "comment-incentives": false,
   "register-wallet-with-verification": false,
+  "openai-api-key": "",
```
[the remark this answers is not in the document]
Comment (Keyrxng) [reply ubiquibot-config-default.json:11]: If this is not set or is an empty string then the OpenAI API will throw an error by default anyway

**Comment by pavlovcik** [comment]
It could be interesting to plug this into CI on push so that it won't pass until the bot thinks it passed. Until we implement automatically closing pull requests due to poor performance though, we should not automatically invoke the `/review`.

File: src/helpers/gpt.ts (hunk 10 above)
Comment (Keyrxng) [reply src/helpers/gpt.ts:257]: It seemed to help a lot with being less anal and pernickety about like white space in the gitignore and those silly requests

File: src/handlers/comment/commands.ts (hunk 11 above)
Comment (Keyrxng) [reply src/handlers/comment/commands.ts:9]: Choose this one because, at least I do not have GPT-4 api access idk if that's a thing but I sure don't have access through my sub. 

This also has the fattest token limit out of the available models excluding the 32k models.

File: ubiquibot-config-default.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (Keyrxng) [reply ubiquibot-config-default.json:11]: - https://github.com/Keyrxng/UbiquityBotTesting/issues/44#issuecomment-1718081294

File: src/helpers/gpt.ts (hunk 10 above)
Comment (pavlovcik) [reply src/helpers/gpt.ts:257]: I'm a bit skeptical about that but we'll roll with your current change.

File: src/handlers/comment/commands.ts (hunk 11 above)
Comment (pavlovcik) [reply src/handlers/comment/commands.ts:9]: It is double the price per token though but lets worry about cost optimization later.

Comment (Keyrxng) [reply src/handlers/comment/commands.ts:9]: Yeah that's the fun part, tiktoken is the best token library for dealing with that but it's a python library. I found a ts wrapper for it but haven't used it

File: ubiquibot-config-default.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (pavlovcik) [reply ubiquibot-config-default.json:11]: Can you change the error to be like? 

```diff
! start with a bang
```

> If this is not set or is an empty string then the OpenAI API will throw an error by default anyway, should I still implement null and throw our own error?

It would provide clarity to our early users if we explain how to enable this feature. Basically it should say something like

"You must configure the `openai-api-key` property in the bot configuration in order to use AI powered features."

[the remark this answers is not in the document]
Comment (pavlovcik) [reply ubiquibot-config-default.json:11]: - #770

[the remark this answers is not in the document]
Comment (Keyrxng) [reply ubiquibot-config-default.json:11]: - https://github.com/Keyrxng/UbiquityBotTesting/pull/46#issuecomment-1719861580

File: src/helpers/gpt.ts (hunk 8 above)
Comment (Keyrxng) [reply src/helpers/gpt.ts]: I feel this can be resolved now.

**Comment by 0xcodercrane** [comment]
> It could be interesting to plug this into CI on push so that it won't pass until the bot thinks it passed. Until we implement automatically closing pull requests due to poor performance though, we should not automatically invoke the `/review`.

good point. we can create an ai-review ci/cd running on each pull request.

**Comment by pavlovcik** [comment]
I'm really looking forward for this one to get merged in because I think it will be able to help us merge the other pull requests faster. @Keyrxng any updates?

File: src/configs/shared.ts
Comment (pavlovcik) [inline src/configs/shared.ts:1]: @0xcodercrane shouldn't these all be moved to `ubiquibot-config-default.ts` ?

**Comment by Keyrxng** [comment]
> any updates?

I felt that I was waiting for you guys but I will stop work on Blame and just ensure everything is operating inline with the favourite you mentioned which I'm sure it is. So long as it is my work with this for now is done unless there are further requests. Will let you know in an a couple hours.

**Comment by Keyrxng** [comment]
- https://github.com/Keyrxng/UbiquityBotTesting/pull/54#issuecomment-1728451567 failed spec with the rest following having 'passed'. Couple of slight errors missed like missing a closing tag but caught others, but this isn't meant to be 100% accurate is it at least not right away?

- https://github.com/Keyrxng/UbiquityBotTesting/pull/41#issuecomment-1728476850 failed spec, again, this pr is throwing a lot of false positives with some checks it understands UUPS and others it doesn't, so some pass while others fail despite calling it concurrently.

- https://github.com/Keyrxng/UbiquityBotTesting/pull/46#issuecomment-1728431464 big spec achieved

- https://github.com/Keyrxng/UbiquityBotTesting/pull/46#issuecomment-1728408458 big spec failed

So the response is short and sweet if it's meeting the spec while when it's failing it's going into depth with things and suggesting how to meet the spec or what is missing from the current implementation.

Done my best to remove any type of overview or listings, I think that ratio you mentioned is better in the more recent stuff than the one you said was your favourite so far.

**Comment by Keyrxng** [comment]
What would be so ideal is if it could be tested on closed prs or something as that amount of real data would be so much better to work with than my shitty gpt-assisted test issues 🤣 Hope it's looking better tho, lmk and I'll wrap this up tomorrow.

File: src/configs/shared.ts
Comment (0xcodercrane) [reply src/configs/shared.ts:1]: https://github.com/ubiquity/ubiquibot/issues/790

File: ubiquibot-config-default.json [hunk 16]
```diff
@@ -8,6 +8,8 @@
   "disable-analytics": false,
   "comment-incentives": false,
   "register-wallet-with-verification": false,
+  "openai-api-key": null,
+  "openai-token-limit": 8000,
```
Comment (pavlovcik) [inline ubiquibot-config-default.json:12]: If you're using a 16k model, shouldn't the default be 16k `max_tokens` ?

**Comment by pavlovcik** [comment]
Does it make sense to eventually turn all the AI functions into `-background` functions? 

https://docs.netlify.com/functions/background-functions/

File: ubiquibot-config-default.json [hunk 17]
```diff
@@ -8,6 +8,8 @@
   "disable-analytics": false,
   "comment-incentives": false,
   "register-wallet-with-verification": false,
+  "openai-api-key": null,
+  "openai-max-tokens": 8000,
```
Comment (pavlovcik) [inline ubiquibot-config-default.json:12]: This is still 8000

**Review by 0x4007 — CHANGES_REQUESTED** [review]

**Comment by 0xcodercrane** [comment]
> Does it make sense to eventually turn all the AI functions into `-background` functions?
> 
> https://docs.netlify.com/functions/background-functions/
> 
> @0xcodercrane rfc

The background function is a new feature of netlify and I think it's worth to take the 15 mins execution timeout benefit of that feature. but my concern from their docs is the pricing plan. they just say it may not be available on all the pricing plans.

File: ubiquibot-config-default.json (hunk 17 above)
Comment (cuddle-dev) [reply ubiquibot-config-default.json:12]: 1/ Why put the null value in default.json here?
2/ @pavlovcik What's your recommendation value? If you put it in the comment, I can directly push to speed up the merge.

**Comment by Keyrxng** [comment]
Okay I'm back in the game

All good to pull things from this pr and start with a clean commit history from the most recent dev branch, seems cleaner and I can avoid dealing with all of the conflicts?
